apps/maestro/src/lib/hooks/useSwitchChain.ts

```diff
@@ -1,31 +1,78 @@
+
+import { useConnectWallet, useWallets } from "@mysten/dapp-kit";
+import { useWeb3Modal } from "@web3modal/wagmi/react";
 import { useSwitchChain as useWagmiSwitchChain } from "wagmi";
 
-import { NEXT_PUBLIC_NETWORK_ENV } from "~/config/env";
+import { suiChainConfig } from "~/config/chains";
 import { useChainId, useDisconnect } from "~/lib/hooks";
-
-// TODO: fixing this once we have the sui chain data
-const CHAIN_ID_SUI = NEXT_PUBLIC_NETWORK_ENV === "mainnet" ? 101 : 103;
+import { isValidEVMAddress } from "../utils/validation";
 
 export function useSwitchChain() {
   const { switchChain: switchChainWagmi } = useWagmiSwitchChain();
   const { disconnect } = useDisconnect();
   const currentChainId = useChainId();
+  const wallets = useWallets();
+  const { mutate: connect } = useConnectWallet();
+  const { open: openWeb3Modal } = useWeb3Modal();
+  const { data: sessionData } = useSession();
+  const [pendingChainId, setPendingChainId] = useLocalStorageState<
+    number | null
+  >("@maestro/pending-chain-id", null);
+
+  useEffect(() => {
+    // wait until the user has connected their wallet and has a pending chain id before switching
+    if (
+      sessionData?.address.length &&
+      isValidEVMAddress(sessionData?.address) &&
+      pendingChainId
+    ) {
+      switchChainWagmi({ chainId: pendingChainId });
+      setPendingChainId(null);
+    }
+  }, [sessionData, pendingChainId, switchChainWagmi, setPendingChainId]);
 
-  function switchChain({ chainId }: { chainId: number }) {
-    const isTargetChainSui = chainId === CHAIN_ID_SUI;
-    const isCurrentChainSui = currentChainId === CHAIN_ID_SUI;
-
-    if (isTargetChainSui && !isCurrentChainSui) {
-      disconnect();
-    } else if (!isTargetChainSui && isCurrentChainSui) {
-      disconnect();
-      switchChainWagmi({ chainId });
-    } else if (!isTargetChainSui && !isCurrentChainSui) {
-      switchChainWagmi({ chainId });
+  const tryConnectSuiWallet = () => {
+    for (const wallet of wallets) {
+      try {
+        // Attempt to connect to each wallet
+        connect({ wallet });
+        return true;
+      } catch (error) {
+        continue;
+      }
+    }
+    return false;
+  };
+
+  async function switchChain({ chainId }: { chainId: number }) {
+    if (chainId) {
+      const isTargetChainSui = chainId === suiChainConfig.id;
+      const isCurrentChainSui = currentChainId === suiChainConfig.id;
+      const evmToEvm = !isTargetChainSui && !isCurrentChainSui;
+      const evmToSui = isTargetChainSui && !isCurrentChainSui;
+      const suiToEvm = !isTargetChainSui && isCurrentChainSui;
+      if (evmToSui) {
+        // EVM to Sui
+        disconnect();
+        tryConnectSuiWallet();
+      } else if (suiToEvm) {
+        // Sui to EVM
+        disconnect();
+        setPendingChainId(chainId);
+        await openWeb3Modal();
+      } else if (evmToEvm && chainId) {
+        // EVM to EVM
+        switchChainWagmi({ chainId });
+      }
```

Comment:
Prepare for the growing chains supported here. Create more robust structure to support additional chains easily. 

My random ideas are similar to the following, feel free to adapt it:

```ts
type ChainType = "evm" | "sui"

interface ChainConfig {
 id: number
 type: ChainType
}

interface ChainConnector {
 connect()
 disconnect() 
 switchChain()
}

class EVMConnector implements ChainConnector {
 connect() {}
 disconnect() {}
 switchChain() {}
}

class SuiConnector implements ChainConnector {
 connect() {}
 disconnect() {}
 switchChain() {}
}
```